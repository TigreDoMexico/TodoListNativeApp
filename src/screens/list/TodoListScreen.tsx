import React, { Component, createRef, useRef, useState } from 'react';
import { FlatList, StyleSheet, Text, View } from 'react-native';
import { useSelector, useDispatch } from 'react-redux';

import iTask from '../../model/interface/iTask';
import Task from '../../model/task';
import NewTaskInput from '../../ui/components/creationInput/NewTaskInput';
import TaskList from '../../ui/components/list/TaskList';
import EditTaskModal from '../../ui/components/modal/EditTaskModal';

import { iTaskListReducer } from '../../store/reducers/TaskListReducer/types';

import Screen from '../Screen';
import { addNewTask, changeTaskStatus, deleteTask, updateTask } from '../../store/reducers/TaskListReducer';
import { ModalHandles } from '../../ui/components/modal/EditTaskModal/types';

const TodoListScreen = () => {
    const editModalRef = createRef<ModalHandles>();
    
    const taskList = useSelector((state: { TaskListState: iTaskListReducer }) =>
        state.TaskListState.tasks)

    const dispatch = useDispatch()

    const onChangeTaskStatusHandler = (taskId: string) => dispatch(changeTaskStatus(taskId))    

    const onEditTaskHandler = (taskId: string) => {
        editModalRef.current?.openModal();
        editModalRef.current?.setTask(taskList.filter(el => el.id === taskId)[0]);
    }

    const onUpdateTask = (task: iTask) => dispatch(updateTask(task));
    
    const onDeleteTask = (taskId: string) => dispatch(deleteTask(taskId))

    const onSaveTaskHandler = (taskName: string) => {
        var task = new Task(taskName)
        dispatch(addNewTask(task))
    }

    return (
        <Screen pageTitle="Lista de Tarefas">
            <View style={styles.addTaskContainer}>
                <NewTaskInput placeholder="Escreva a tarefa aqui" onSaveTask={onSaveTaskHandler} />
            </View>
            <View style={styles.taskListContainer}>
                <TaskList
                    tasks={taskList}
                    onChangeTaskStatus={onChangeTaskStatusHandler}
                    onEdit={onEditTaskHandler}
                />
            </View>            
            <EditTaskModal
                ref={editModalRef}
                onSaveTask={onUpdateTask}
                onDeleteTask={onDeleteTask}
            />
        </Screen>
    );
}

const styles = StyleSheet.create({
    addTaskContainer: {
        marginTop: 20,
        marginBottom: 30,
        marginHorizontal: 20
    },
    taskListContainer: {
        flex: 1,
        marginHorizontal: 20,
        marginBottom: 10,
    }
});

export default TodoListScreen;