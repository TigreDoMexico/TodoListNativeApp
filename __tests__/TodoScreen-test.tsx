import React from 'react'
import {Provider} from 'react-redux'
import renderer from 'react-test-renderer'

import configureStore from '../src/store/store'

import TodoListScreen from '../src/screens/list/TodoListScreen';

//TODO -> ERROR Cannot read property 'task' of undefined
const tree = renderer.create(
    <Provider store={configureStore}>
        <TodoListScreen />
    </Provider>
);

test('Criando uma tela de tarefas - Esperado combinar com o Snapshot',() => {
    expect(tree).toMatchSnapshot();
})

describe("Verificando se o Store vem com array vazio", () => {
    it("Precisa vir com o Array de Tasvazio", () => {
        expect(configureStore.getState().TaskListReducer.tasks.length).toEqual(0)
    })
})






