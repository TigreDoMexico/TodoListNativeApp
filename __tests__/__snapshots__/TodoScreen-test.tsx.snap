// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Criando uma tela de tarefas - Esperado combinar com o Snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FCF8EC",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#79A3B1",
        "height": 70,
        "justifyContent": "center",
        "paddingTop": 25,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FCF8EC",
          "fontSize": 25,
          "fontWeight": "bold",
        }
      }
    >
      Lista de Tarefas
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "marginBottom": 30,
          "marginHorizontal": 20,
          "marginTop": 20,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-around",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          placeholder="Escreva a tarefa aqui"
          rejectResponderTermination={true}
          style={
            Object {
              "borderColor": "#D0E8F2",
              "borderRadius": 10,
              "borderWidth": 2,
              "flex": 1,
              "fontSize": 16,
              "paddingHorizontal": 20,
              "paddingVertical": 10,
            }
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          style={
            Object {
              "marginLeft": 10,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#456268",
                  "borderRadius": 5,
                  "paddingHorizontal": 18,
                  "paddingVertical": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#FCF8EC",
                    "fontSize": 15,
                  }
                }
              >
                Adicionar
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginBottom": 10,
          "marginHorizontal": 20,
        }
      }
    >
      <RCTScrollView
        ItemSeparatorComponent={[Function]}
        data={Array []}
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View />
      </RCTScrollView>
    </View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Modal
        animationType="none"
        deviceHeight={null}
        deviceWidth={null}
        hardwareAccelerated={false}
        hideModalContentWhileAnimating={false}
        onBackdropPress={[Function]}
        onModalHide={[Function]}
        onModalWillHide={[Function]}
        onModalWillShow={[Function]}
        onRequestClose={[Function]}
        scrollHorizontal={false}
        scrollOffset={0}
        scrollOffsetMax={0}
        scrollTo={null}
        supportedOrientations={
          Array [
            "portrait",
            "landscape",
          ]
        }
        swipeThreshold={100}
        transparent={true}
        visible={false}
      >
        <View
          accessible={true}
          focusable={true}
          forwardedRef={[Function]}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "black",
              "bottom": 0,
              "height": 1334,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 750,
            }
          }
        />
        <View
          onLayout={[Function]}
          pointerEvents="box-none"
          style={
            Array [
              Array [
                Object {
                  "margin": 37.5,
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                  ],
                },
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
                Object {
                  "justifyContent": "flex-end",
                  "margin": 0,
                },
                Object {
                  "margin": 0,
                },
              ],
              Object {
                "paddingBottom": 0,
              },
            ]
          }
        >
          <View
            deviceHeight={null}
            deviceWidth={null}
            forwardedRef={[Function]}
            hideModalContentWhileAnimating={false}
            onBackdropPress={[Function]}
            onModalHide={[Function]}
            onModalWillHide={[Function]}
            onModalWillShow={[Function]}
            pointerEvents="box-none"
            scrollHorizontal={false}
            scrollOffset={0}
            scrollOffsetMax={0}
            scrollTo={null}
            style={
              Object {
                "flex": 1,
                "justifyContent": "flex-end",
                "margin": 0,
                "transform": Array [
                  Object {
                    "translateY": 0,
                  },
                ],
              }
            }
            supportedOrientations={
              Array [
                "portrait",
                "landscape",
              ]
            }
            swipeThreshold={100}
          >
            <View
              style={
                Object {
                  "backgroundColor": "#FCF8EC",
                  "height": "50%",
                  "paddingBottom": 20,
                  "paddingHorizontal": 20,
                  "paddingTop": 30,
                }
              }
            >
              <View>
                <Text
                  style={
                    Object {
                      "alignItems": "center",
                      "color": "#456268",
                      "fontSize": 20,
                      "justifyContent": "center",
                    }
                  }
                >
                  Edição da Tarefa
                </Text>
              </View>
              <View
                style={
                  Object {
                    "height": 50,
                    "marginBottom": 30,
                    "marginTop": 20,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#F35958",
                    }
                  }
                >
                  A tarefa não pode ter um nome vazio
                </Text>
                <View
                  style={
                    Object {
                      "height": 50,
                    }
                  }
                >
                  <TextInput
                    allowFontScaling={true}
                    onChangeText={[Function]}
                    placeholder="Nome da tarefa"
                    rejectResponderTermination={true}
                    style={
                      Object {
                        "borderColor": "#D0E8F2",
                        "borderRadius": 10,
                        "borderWidth": 2,
                        "flex": 1,
                        "fontSize": 16,
                        "paddingHorizontal": 20,
                        "paddingVertical": 10,
                      }
                    }
                    underlineColorAndroid="transparent"
                    value=""
                  />
                </View>
              </View>
              <View
                style={
                  Object {
                    "marginBottom": 20,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#456268",
                      "fontSize": 15,
                      "marginBottom": 10,
                    }
                  }
                >
                  Cor da Tarefa
                </Text>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <View
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "height": 40,
                            "width": 60,
                          },
                          Object {
                            "backgroundColor": "#79A3B1",
                          },
                        ]
                      }
                    />
                  </View>
                  <View
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "height": 40,
                            "width": 60,
                          },
                          Object {
                            "backgroundColor": "#FF7F50",
                          },
                        ]
                      }
                    />
                  </View>
                  <View
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "height": 40,
                            "width": 60,
                          },
                          Object {
                            "backgroundColor": "#79B187",
                          },
                        ]
                      }
                    />
                  </View>
                  <View
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "height": 40,
                            "width": 60,
                          },
                          Object {
                            "backgroundColor": "#CA4fCA",
                          },
                        ]
                      }
                    />
                  </View>
                  <View
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "height": 40,
                            "width": 60,
                          },
                          Object {
                            "backgroundColor": "#FF8684",
                          },
                        ]
                      }
                    />
                  </View>
                </View>
                <Text>
                  <Text>
                    
                  </Text>
                  <Text>
                    
                  </Text>
                  <Text />
                </Text>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "bottom": 0,
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                  }
                }
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "opacity": 1,
                    }
                  }
                  testID="touchable_custom_button"
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "justifyContent": "center",
                        },
                        Object {
                          "backgroundColor": "#F35958",
                          "paddingHorizontal": 30,
                          "paddingVertical": 10,
                          "width": 375,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Object {
                          "flexDirection": "row",
                        }
                      }
                    >
                      <Text
                        style={
                          Object {
                            "color": "#FCF8EC",
                            "fontSize": 18,
                            "marginRight": 10,
                          }
                        }
                      >
                        Excluir
                      </Text>
                      <Text />
                    </View>
                  </View>
                </View>
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "opacity": 1,
                    }
                  }
                  testID="touchable_custom_button"
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "justifyContent": "center",
                        },
                        Object {
                          "backgroundColor": "#16A082",
                          "paddingHorizontal": 30,
                          "paddingVertical": 10,
                          "width": 375,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Object {
                          "flexDirection": "row",
                        }
                      }
                    >
                      <Text
                        style={
                          Object {
                            "color": "#FCF8EC",
                            "fontSize": 18,
                            "marginRight": 10,
                          }
                        }
                      >
                        Salvar
                      </Text>
                      <Text />
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </Modal>
    </View>
  </View>
</View>
`;
